src/icon.rs

```diff
@@ -12,108 +15,121 @@ pub(crate) struct Pixel {
 
 pub(crate) const PIXEL_SIZE: usize = mem::size_of::<Pixel>();
 
-#[derive(Debug)]
-/// An error produced when using `Icon::from_rgba` with invalid arguments.
-pub enum BadIcon {
-    /// Produced when the length of the `rgba` argument isn't divisible by 4, thus `rgba` can't be
-    /// safely interpreted as 32bpp RGBA pixels.
-    ByteCountNotDivisibleBy4 { byte_count: usize },
-    /// Produced when the number of pixels (`rgba.len() / 4`) isn't equal to `width * height`.
-    /// At least one of your arguments is incorrect.
-    DimensionsVsPixelCount {
-        width: u32,
-        height: u32,
-        width_x_height: usize,
-        pixel_count: usize,
-    },
-    /// Produced when underlying OS functionality failed to create the icon
-    OsError(io::Error),
-}
-
-impl fmt::Display for BadIcon {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            BadIcon::ByteCountNotDivisibleBy4 { byte_count } => write!(f,
-                "The length of the `rgba` argument ({:?}) isn't divisible by 4, making it impossible to interpret as 32bpp RGBA pixels.",
-                byte_count,
-            ),
-            BadIcon::DimensionsVsPixelCount {
-                width,
-                height,
-                width_x_height,
-                pixel_count,
-            } => write!(f,
-                "The specified dimensions ({:?}x{:?}) don't match the number of pixels supplied by the `rgba` argument ({:?}). For those dimensions, the expected pixel count is {:?}.",
-                width, height, pixel_count, width_x_height,
-            ),
-            BadIcon::OsError(e) => write!(f, "OS error when instantiating the icon: {:?}", e),
-        }
-    }
-}
-
-impl Error for BadIcon {
-    fn source(&self) -> Option<&(dyn Error + 'static)> {
-        Some(self)
-    }
-}
-
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub(crate) struct RgbaIcon {
-    pub(crate) rgba: Vec<u8>,
-    pub(crate) width: u32,
-    pub(crate) height: u32,
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct RgbaIcon<I: Deref<Target = [u8]>> {
+    pub(crate) rgba: I,
+    pub(crate) size: PhysicalSize<u32>,
+    pub(crate) hot_spot: PhysicalPosition<u32>,
```

Comment:
I don't really like that we have a cursor and normal icons under the same `RgbaIcon`, in general, we can name it like `RgbaBuffer`, since the intent to have `RgbaBuffer` and have `Icon` and a new type for cursor icons, which will use that `RgbaBuffer` under the hood?

It's just strange that we have a field that shouldn't be used in any ways in certain scenarios, but in other cases it's fine to use them. Such issues should be solved on type system level.

Reply:
That's a decent point. I've updated the API to split the custom window type and custom cursor type into two parts, though it's not documented yet. Could you take a look at it?